src/app/admin/edit-product/edit-product.component.ts

```diff
@@ -0,0 +1,87 @@
+import {
+  ChangeDetectionStrategy,
+  Component,
+  OnDestroy,
+  OnInit,
+} from '@angular/core';
+import { FormBuilder, FormGroup, Validators } from '@angular/forms';
+import { ActivatedRoute, Router } from '@angular/router';
+
+import { BehaviorSubject, of, Subject } from 'rxjs';
+import { switchMap, takeUntil } from 'rxjs/operators';
+
+import { Product } from '../../products/product.interface';
+import { ProductsService } from '../../products/products.service';
+import { NotificationService } from '../../core/notification.service';
+import { HttpErrorResponse } from '@angular/common/http';
+
+@Component({
+  selector: 'app-edit-product',
+  templateUrl: './edit-product.component.html',
+  styleUrls: ['./edit-product.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class EditProductComponent implements OnInit, OnDestroy {
+  form$: BehaviorSubject<FormGroup | null> =
+    new BehaviorSubject<FormGroup | null>(null);
+  product: Product | null = null;
+  requestInProgress = false;
+
+  private readonly onDestroy$: Subject<void> = new Subject();
+
+  constructor(
+    private readonly activatedRoute: ActivatedRoute,
+    private readonly fb: FormBuilder,
+    private readonly notificationService: NotificationService,
+    private readonly productsService: ProductsService,
+    private readonly router: Router
+  ) {}
+
+  ngOnInit(): void {
+    this.activatedRoute.paramMap
+      .pipe(
+        switchMap((paramMap) =>
+          paramMap.has('productId')
+            ? this.productsService.getProductById(paramMap.get('productId')!)
+            : of(null)
+        ),
+        takeUntil(this.onDestroy$)
+      )
+      .subscribe((product) => {
+        this.product = product;
+        this.buildForm();
+      });
+  }
+
+  ngOnDestroy(): void {
+    this.onDestroy$.next();
+    this.onDestroy$.complete();
+  }
+
+  editProduct(): void {
+    const product: Product = this.form$.value!.value;
+    const editProduct$ = this.product
+      ? this.productsService.editProduct(this.product.id, product)
+      : this.productsService.createNewProduct(product);
+
+    this.requestInProgress = true;
+    editProduct$.subscribe(
+      () => this.router.navigate(['../'], { relativeTo: this.activatedRoute }),
+      (error: HttpErrorResponse) => {
```

Comment:
unknown

Reply:
what's wrong with HttpErrorResponse? Using unknown looks rather ugly to me

Reply:
Because you can never be sure what error type you get there. Following this approach might cause error in the error handling clause.
It makes it look uglier, but it makes the code more reliable and thus it's better.

Reply:
It's safer not to type assert (`error as string`) but to make checks for the error type. Only this way you can be sure in the production that your code will work as expected.

Reply:
Rolled back to unknown. About the notification, maybe the best way is to remove it? Those type checks you mentioned strongly depend on the backend api, which is probably out of scope of this angular shop template. What do you think?

Reply:
Not sure I understand you